src/DeepLinkModule/javasource/deeplink/actions/StartDeeplinkJava.java

```diff
@@ -200,28 +203,30 @@ else if (!deeplink.getAllowGuests().booleanValue() && session.getUser().isAnonym
 			
 			//switch to the users context
 			context = session.createContext(); 
-			IUser sessionUserObj = session.getUser();
+			IUser sessionUserObj = session.getUser(sessionContext);
 			String userName = sessionUserObj.getName();
 			
 			// Retrieve the language set for the deeplink.
 			Language language = deeplink.getDeepLink_Language();
 			if (language != null) {
 				// Deeplink has a language, put it on the session user.
-				User user = User.initialize(context, sessionUserObj.getMendixObject());
+				User user = User.initialize(sessionContext, sessionUserObj.getMendixObject());
 				user.setUser_Language(language);
 				user.commit();
 			}
 
 			//we have a valid session, further security is enforced by the application
 			
 			//first, remove old pending links
-			List<IMendixObject> pendinglinks = Core.retrieveXPathQueryEscaped(context, "//%s[%s='%s' and %s='%s']",
-					PendingLink.getType(), 
-					PendingLink.MemberNames.PendingLink_DeepLink.toString(), String.valueOf(deeplink.getMendixObject().getId().toLong()),
-					PendingLink.MemberNames.User.toString(), userName 
-			);
+			List<IMendixObject> pendinglinks = Core.createXPathQuery("//$type[$linkMember='$mxId' and $userMember='$username']")

```

Comment:
I'm afraid setting entity name and attributes through a variable won't work. We may need to check how much `$type`, `$linkMember` and `$userMember` are vulnerable to injections and protect from it in some other way if it is the case.

`$mxId` and `$username` can still be a variable, so we're covered for them.